design_docs/parallel_tasks.md

```diff
@@ -0,0 +1,396 @@
+the failure of one analysis task not to interrupt execution of other analyses.
+
+In a future PR, this work can be expanded to include checking if the appropriate
+analysis member (AM) was turned on during the run and skipping any analysis tasks that
+depend on that AM if not (Issue #58).
+
+<h4> Design solution: Lock files written by multiple tasks <br>
+Date last modified: 2017/03/10 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+
+Teh design solution is based on the process lock in the fasteners package: 
+http://fasteners.readthedocs.io/en/latest/examples.html#interprocess-locks
+
+Currently, only mapping files should be written by multiple tasks, requiring locks.
+
+The algorithm consists of 2 changes.  First, I removed the option `overwriteMappingFiles`,
+which is now always `False`---if a mapping file exists, it is not overwritten.  This
+was necessary because now only one task will write a given mapping file if it doesn't
+already exist and the other tasks will wait for it to be written.  Then, all tasks
+know there is a valid mapping file that they can read without having to lock the file.
+
+The second change was to add a lock around the subprocess call to `ESMF_RegridWeightGen`
+that make sure only one process generates the mapping file.  Each process attempts to
+acquire the lock and checks if the mapping file already exists once it acquires the
+lock.  If not, it generates the mapping file and releases the lock.  If so, it just
+releases the lock and moves on.  Thus, only the first process to acquire the lock
+generates the mapping file and the others wait until it is finished.
+
+<h4> Design solution: Task parallelism should work on either login or compute nodes <br>
+Date last modified: 2017/02/23 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+For the time being, I propose to address only task parallelism on the login nodes and to
+extend the parallelism to work robustly on compute nodes as a separate project.
+Nevertheless, I will seek to implement this design in a way that should be conducive to
+this later extension.  Likely what will be required is a robust way of adding a prefix
+to the commandline (e.g. `aprun -np 1`) when calling subprocesses.  Adding such a prefix
+should be relatively simple.
+
+<h4> Design solution: There may need to be a way to limit the memory used by a task <br>
+Date last modified: 2017/02/23 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+I am not very familiar with `dask` within `xarray` and I do not intend to address this
+consideration directly in this project.  However, on my brief investigation, it seems like
+the proper way to handle this may be to have a `chunk` config option either for all tasks
+or for individual tasks that can be used to control the size of data in memory. I think
+such an approach can be investigated in parallel to this project.  An intermediate solution
+for situations where memory is limited would be to set `parallelTaskCount` to a small number.
```

Comment:
I think we just keep an eye on this for now and see if problems arise.  It isn't clear to me whether we will, at least in the short term, for our use cases.  If not, awesome!  If we do have problems I'm obviously happy to help.  I'm guessing we will be ok because we aren't typically doing anything too compute-heavy because this is done in-situ via MPAS-O analysis members.